File: scala/core/src/main/scala/blockchain/crypto/KesProduct.scala
package blockchain.crypto

import blockchain.crypto.KesBinaryTree.*
import com.google.common.primitives.Longs

import java.security.SecureRandom
import scala.collection.mutable

class KesProduct extends ProductComposition {

  def createKeyPair(
      seed: Array[Byte],
      height: (Int, Int),
      offset: Long
  ): (SecretKeyKesProduct, VerificationKeyKesProduct) = {
    val sk = generateSecretKey(seed, height._1, height._2)
    val pk = generateVerificationKey(sk)
    (
      SecretKeyKesProduct(
        sk._1,
        sk._2,
        sk._3,
        SignatureKesSum(
          sk._4._1,
          sk._4._2,
          sk._4._3
        ),
        offset
      ),
      VerificationKeyKesProduct(pk._1, pk._2)
    )
  }

  def sign(
      privateKey: SecretKeyKesProduct,
      message: Array[Byte]
  ): SignatureKesProduct = {
    val prodSig = sign(unpackSecret(privateKey), message)

    SignatureKesProduct(
      SignatureKesSum(
        prodSig._1._1,
        prodSig._1._2,
        prodSig._1._3
      ),
      SignatureKesSum(
        prodSig._2._1,
        prodSig._2._2,
        prodSig._2._3
      ),
      prodSig._3
    )
  }

  def verify(
      signature: SignatureKesProduct,
      message: Array[Byte],
      verifyKey: VerificationKeyKesProduct
  ): Boolean = {
    val prodSig = (
      (
        signature.superSignature.verificationKey,
        signature.superSignature.signature,
        signature.superSignature.witness.toVector
      ),
      (
        signature.subSignature.verificationKey,
        signature.subSignature.signature,
        signature.subSignature.witness.toVector
      ),
      signature.subRoot
    )

    val sumVk = (verifyKey.value, verifyKey.step)
    verify(prodSig, message.toArray, sumVk)
  }

  def update(
      privateKey: SecretKeyKesProduct,
      steps: Int
  ): SecretKeyKesProduct = {
    val sk = updateKey(unpackSecret(privateKey), steps)

    SecretKeyKesProduct(
      sk._1,
      sk._2,
      sk._3,
      SignatureKesSum(
        sk._4._1,
        sk._4._2,
        sk._4._3
      ),
      privateKey.offset
    )
  }

  def getCurrentStep(privateKey: SecretKeyKesProduct): Int = getKeyTime(
    unpackSecret(privateKey)
  )

  def getMaxStep(privateKey: SecretKeyKesProduct): Int = exp(
    getTreeHeight(privateKey.superTree) + getTreeHeight(privateKey.subTree)
  )

  def getVerificationKey(
      privateKey: SecretKeyKesProduct
  ): VerificationKeyKesProduct = {
    val vk = generateVerificationKey(unpackSecret(privateKey))
    VerificationKeyKesProduct(vk._1, vk._2)
  }

  private def unpackSecret(privateKey: SecretKeyKesProduct): SK =
    (
      privateKey.superTree,
      privateKey.subTree,
      privateKey.nextSubSeed,
      (
        privateKey.subSignature.verificationKey,
        privateKey.subSignature.signature,
        privateKey.subSignature.witness.toVector
      )
    )

}

case class SecretKeyKesProduct(
    superTree: KesBinaryTree, // Hour hand
    subTree: KesBinaryTree, // Minute hand
    nextSubSeed: Array[Byte],
    subSignature: SignatureKesSum,
    offset: Long
) {

  override def equals(obj: Any): Boolean =
    obj.asInstanceOf[Matchable] match {
      case s: SecretKeyKesProduct =>
        superTree == s.superTree &&
        subTree == s.subTree &&
        java.util.Arrays.equals(nextSubSeed, s.nextSubSeed) &&
        subSignature == s.subSignature &&
        offset == s.offset
      case _ =>
        false
    }

  override def hashCode(): Int = {
    var r = 1

    r = 31 * r + superTree.hashCode()
    r = 31 * r + subTree.hashCode()
    r = 31 * r + java.util.Arrays.hashCode(nextSubSeed)
    r = 31 * r + subSignature.hashCode()
    r = 31 * r + offset.hashCode()
    r
  }

  def toByteArray: Array[Byte] =
    mutable.ArrayBuilder
      .make[Byte]
      .addAll(superTree.toByteArray)
      .addAll(subTree.toByteArray)
      .addAll(nextSubSeed)
      .addAll(subSignature.toByteArray)
      .addAll(Longs.toByteArray(offset))
      .result()
}

case class SignatureKesProduct(
    superSignature: SignatureKesSum,
    subSignature: SignatureKesSum,
    subRoot: Array[Byte]
) {

  override def hashCode(): Int = {
    var r = 1
    r = 31 * r + superSignature.hashCode()
    r = 31 * r + subSignature.hashCode()
    r = 31 * r + java.util.Arrays.hashCode(subRoot)
    r
  }

  override def equals(other: Any): Boolean =
    other.asInstanceOf[Matchable] match {
      case kesProduct: SignatureKesProduct =>
        superSignature == kesProduct.superSignature &&
        subSignature == kesProduct.subSignature &&
        subRoot.sameElements(kesProduct.subRoot)

      case _ => false
    }
}

case class VerificationKeyKesProduct(value: Array[Byte], step: Int) {

  override def hashCode(): Int = {
    var r = 1
    r = 31 * r + java.util.Arrays.hashCode(value)
    r = 31 * r + step.hashCode
    r
  }

  override def equals(other: Any): Boolean =
    other.asInstanceOf[Matchable] match {
      case vk: VerificationKeyKesProduct =>
        value.sameElements(vk.value) &&
        step == vk.step
      case _ => false
    }
}

/** Credit to Aaron Schutza
  */
private[crypto] class ProductComposition extends KesEd25519Blake2b256 {

  protected val sumComposition = new SumComposition

  protected val random: SecureRandom = new SecureRandom

  override type SIG = (sumComposition.SIG, sumComposition.SIG, Array[Byte])
  override type VK = (Array[Byte], Int)
  override type SK =
    (sumComposition.SK, sumComposition.SK, Array[Byte], sumComposition.SIG)

  /** Get the current time step of an MMM key
    * @param key
    *   MMM key to be inspected
    * @return
    *   Current time step of key
    */
  private[crypto] def getKeyTime(key: SK): Int = {
    val numSubSteps = exp(sumComposition.getTreeHeight(key._2))
    val tSup = sumComposition.getKeyTime(key._1)
    val tSub = sumComposition.getKeyTime(key._2)
    (tSup * numSubSteps) + tSub
  }

  /** @param key
    * @return
    */
  private[crypto] def generateVerificationKey(key: SK): VK = key._1 match {
    case node: MerkleNode  => (witness(node), getKeyTime(key))
    case leaf: SigningLeaf => (witness(leaf), 0)
    case Empty()           => (Array.fill(hashBytes)(0: Byte), 0)
  }

  /** Generate key in the MMM composition
    * @param seed
    *   input entropy for key generation
    * @return
    */
  private[crypto] def generateSecretKey(
      seed: Array[Byte],
      heightSup: Int,
      heightSub: Int
  ): SK = {
    val rSuper = prng(seed)
    val rSub = prng(rSuper._2)
    val superScheme = sumComposition.generateSecretKey(rSuper._1, heightSup)
    val subScheme = sumComposition.generateSecretKey(rSub._1, heightSub)
    val kesVkSub: sumComposition.VK =
      sumComposition.generateVerificationKey(subScheme)
    val kesSigSup: sumComposition.SIG =
      sumComposition.sign(superScheme, kesVkSub._1)
    random.nextBytes(rSuper._2)
    random.nextBytes(seed)
    (superScheme, subScheme, rSub._2, kesSigSup)
  }

  /** Erases the secret key at the leaf level of a private key in the sum composition Used to commit to a child
    * verification key and then convert the parent private key to a state that can't be used to re-commit to another
    * child key until the next time step
    * @param input
    *   input key
    * @return
    *   new key with overwritten SigningLeaf sk
    */

  private[crypto] def eraseLeafSecretKey(input: KesBinaryTree): KesBinaryTree =
    input match {
      case n: MerkleNode =>
        (n.left, n.right) match {
          case (Empty(), _) =>
            MerkleNode(
              n.seed,
              n.witnessLeft,
              n.witnessRight,
              Empty(),
              eraseLeafSecretKey(n.right)
            )
          case (_, Empty()) =>
            MerkleNode(
              n.seed,
              n.witnessLeft,
              n.witnessRight,
              eraseLeafSecretKey(n.left),
              Empty()
            )
          case (_, _) => throw new Exception("Evolving Key Configuration Error")
        }
      case l: SigningLeaf =>
        random.nextBytes(l.sk)
        SigningLeaf(Array.fill[Byte](sig.SECRET_KEY_SIZE)(0), l.vk)
      case _ => throw new Exception("Evolving Key Configuration Error")
    }

  /** Erases the secret key at the leaf level of a private key in the product composition Used to commit to a child
    * verification key and then convert the parent private key to a state that can't be used to re-commit to another
    * child key until the next time step
    * @param key
    *   input key
    * @return
    *   new key with overwritten child scheme SigningLeaf sk
    */

  private[crypto] def eraseProductLeafSk(key: SK): SK =
    (key._1, eraseLeafSecretKey(key._2), key._3, key._4)

  /** Updates product keys to the specified time step
    * @param key
    *   input key
    * @param step
    *   input desired time step
    * @return
    *   updated key
    */
  private[crypto] def updateKey(key: SK, step: Int): SK = {
    val keyTime = getKeyTime(key)
    val keyTimeSup = sumComposition.getKeyTime(key._1)
    val heightSup = sumComposition.getTreeHeight(key._1)
    val heightSub = sumComposition.getTreeHeight(key._2)
    val totalSteps = exp(heightSup + heightSub)
    val totalStepsSub = exp(heightSub)
    val newKeyTimeSup = step / totalStepsSub
    val newKeyTimeSub = step % totalStepsSub

    def getSeed(
        seeds: (Array[Byte], Array[Byte]),
        iter: Int
    ): (Array[Byte], Array[Byte]) =
      if (iter < newKeyTimeSup) {
        val out = getSeed(prng(seeds._2), iter + 1)
        random.nextBytes(seeds._1)
        random.nextBytes(seeds._2)
        out
      } else seeds

    if (step == 0) key
    else if (step > keyTime && step < totalSteps) {
      if (keyTimeSup < newKeyTimeSup) {
        sumComposition.eraseOldNode(key._2)
        val (s1, s2) = getSeed((Array(), key._3), keyTimeSup)
        val superScheme = sumComposition.evolveKey(key._1, newKeyTimeSup)
        val newSubScheme = sumComposition.generateSecretKey(s1, heightSub)
        random.nextBytes(s1)
        val kesVkSub = sumComposition.generateVerificationKey(newSubScheme)
        val kesSigSuper = sumComposition.sign(superScheme, kesVkSub._1)
        val forwardSecureSuperScheme = eraseLeafSecretKey(superScheme)
        val updatedSubScheme =
          sumComposition.evolveKey(newSubScheme, newKeyTimeSub)
        (forwardSecureSuperScheme, updatedSubScheme, s2, kesSigSuper)
      } else {
        val subScheme = sumComposition.updateKey(key._2, newKeyTimeSub)
        (key._1, subScheme, key._3, key._4)
      }
    } else {
      throw new Error(
        s"Update error - Max steps: $totalSteps, current step: $keyTime, requested increase: $step"
      )
    }
  }

  /** @param key
    * @param m
    * @return
    */
  private[crypto] def sign(key: SK, m: Array[Byte]): SIG =
    (
      key._4,
      sumComposition.sign(key._2, m),
      sumComposition.generateVerificationKey(key._2)._1
    )

  /** Verify MMM signature
    * @param pk
    *   public key of the MMM secret key
    * @param m
    *   message corresponding to signature
    * @param sig
    *   signature to be verified
    * @return
    *   true if signature is valid false if otherwise
    */
  private[crypto] def verify(
      kesSig: SIG,
      m: Array[Byte],
      kesVk: VK
  ): Boolean = {
    val totalStepsSub = exp(kesSig._2._3.length)
    val keyTimeSup = kesVk._2 / totalStepsSub
    val keyTimeSub = kesVk._2 % totalStepsSub

    val verifySup =
      sumComposition.verify(kesSig._1, kesSig._3, (kesVk._1, keyTimeSup))
    val verifySub = sumComposition.verify(kesSig._2, m, (kesSig._3, keyTimeSub))

    verifySup && verifySub
  }
}

File: dart/blockchain/lib/crypto/impl/kes_product.dart
import 'dart:math';
import 'dart:typed_data';

import 'package:blockchain/codecs.dart';
import 'package:blockchain/common/utils.dart';
import 'package:blockchain/crypto/impl/kes_helper.dart';
import 'package:blockchain/crypto/impl/kes_sum.dart';
import 'package:blockchain_protobuf/models/core.pb.dart';
import 'package:fixnum/fixnum.dart';

import '../utils.dart';

abstract class KesProduct {
  Future<KeyPairKesProduct> generateKeyPair(
      List<int> seed, TreeHeight height, Int64 offset);
  Future<SignatureKesProduct> sign(SecretKeyKesProduct sk, List<int> message);
  Future<bool> verify(SignatureKesProduct signature, List<int> message,
      VerificationKeyKesProduct vk);
  Future<SecretKeyKesProduct> update(SecretKeyKesProduct sk, int step);
  Future<int> getCurrentStep(SecretKeyKesProduct sk);
  Future<VerificationKeyKesProduct> generateVerificationKey(
      SecretKeyKesProduct sk);
}

class KesProductImpl extends KesProduct {
  Future<KeyPairKesProduct> generateKeyPair(
      List<int> seed, TreeHeight height, Int64 offset) async {
    final sk = await generateSecretKey(seed, height);
    final vk = await generateVerificationKey(sk);
    return KeyPairKesProduct(sk: sk, vk: vk);
  }

  Future<SignatureKesProduct> sign(
      SecretKeyKesProduct sk, List<int> message) async {
    return SignatureKesProduct()
      ..superSignature = sk.subSignature
      ..subSignature = await kesSum.sign(sk.subTree, message)
      ..subRoot =
          (await kesSum.generateVerificationKey(sk.subTree)).value.base58;
  }

  Future<bool> verify(SignatureKesProduct signature, List<int> message,
      VerificationKeyKesProduct vk) async {
    final totalStepsSub = kesHelper.exp(signature.subSignature.witness.length);
    final keyTimeSup = vk.step ~/ totalStepsSub;
    final keyTimeSub = vk.step % totalStepsSub;
    final superVerification = await kesSum.verify(
        signature.superSignature,
        signature.subRoot.decodeBase58,
        VerificationKeyKesSum(
            value: Uint8List.fromList(vk.value.decodeBase58),
            step: keyTimeSup));
    if (!superVerification) return false;
    final subVerification = await kesSum.verify(
        signature.subSignature,
        message,
        VerificationKeyKesSum(
            value: Uint8List.fromList(signature.subRoot.decodeBase58),
            step: keyTimeSub));
    return subVerification;
  }

  Future<SecretKeyKesProduct> update(SecretKeyKesProduct sk, int step) async {
    if (step == 0) return sk;
    final keyTime = await getCurrentStep(sk);
    final keyTimeSup = await kesSum.getCurrentStep(sk.superTree);
    final heightSup = kesHelper.getTreeHeight(sk.superTree);
    final heightSub = kesHelper.getTreeHeight(sk.subTree);
    final totalSteps = kesHelper.exp(heightSup + heightSub);
    final totalStepsSub = kesHelper.exp(heightSub);
    final newKeyTimeSup = step ~/ totalStepsSub;
    final newKeyTimeSub = step % totalStepsSub;

    Future<(List<int>, List<int>)> getSeed(
        (List<int>, List<int>) seeds, int iter) async {
      if (iter < newKeyTimeSup) {
        final out = getSeed(await kesHelper.prng(seeds.$2), iter + 1);
        kesHelper.overwriteBytes(seeds.$1);
        kesHelper.overwriteBytes(seeds.$2);
        return out;
      } else
        return seeds;
    }

    if (step > keyTime && step < totalSteps) {
      if (keyTimeSup < newKeyTimeSup) {
        kesSum.eraseOldNode(sk.subTree);
        final seeds = await getSeed(([], sk.nextSubSeed), keyTimeSup);
        final superScheme = await kesSum.evolveKey(sk.superTree, newKeyTimeSup);
        final newSubScheme = await kesSum.generateSecretKey(
            Uint8List.fromList(seeds.$1), heightSub);
        kesHelper.overwriteBytes(seeds.$1);
        final kesVkSub = await kesSum.generateVerificationKey(newSubScheme);
        final kesSigSuper = await kesSum.sign(superScheme, kesVkSub.value);
        final forwardSecureSuperScheme = _eraseLeafSecretKey(superScheme);
        final updatedSubScheme =
            await kesSum.evolveKey(newSubScheme, newKeyTimeSub);
        return SecretKeyKesProduct(
          superTree: forwardSecureSuperScheme,
          subTree: updatedSubScheme,
          nextSubSeed: seeds.$2,
          subSignature: kesSigSuper,
          offset: sk.offset, // TODO
        );
      } else {
        final subScheme = await kesSum.update(sk.subTree, newKeyTimeSub);
        return SecretKeyKesProduct(
          superTree: sk.superTree,
          subTree: subScheme,
          nextSubSeed: sk.nextSubSeed,
          subSignature: sk.subSignature,
          offset: sk.offset, // TODO
        );
      }
    }
    throw Exception(
        "Update error - Max steps: $totalSteps, current step: $keyTime, requested increase: $step");
  }

  Future<int> getCurrentStep(SecretKeyKesProduct sk) async {
    final numSubSteps = kesHelper.exp(kesHelper.getTreeHeight(sk.subTree));
    final tSup = await kesSum.getCurrentStep(sk.superTree);
    final tSub = await kesSum.getCurrentStep(sk.subTree);
    return (tSup * numSubSteps) + tSub;
  }

  Future<SecretKeyKesProduct> generateSecretKey(
      List<int> seed, TreeHeight height) async {
    final rSuper = await kesHelper.prng(seed);
    final rSub = await kesHelper.prng(rSuper.$2);
    final superScheme = await kesSum.generateSecretKey(rSuper.$1, height.sup);
    final subScheme = await kesSum.generateSecretKey(rSub.$1, height.sub);
    final kesVkSub = await kesSum.generateVerificationKey(subScheme);
    final kesSigSuper = await kesSum.sign(superScheme, kesVkSub.value);
    kesHelper.overwriteBytes(rSuper.$2);
    kesHelper.overwriteBytes(seed);
    return SecretKeyKesProduct(
      superTree: superScheme,
      subTree: subScheme,
      nextSubSeed: rSub.$2,
      subSignature: kesSigSuper,
      offset: Int64(0), // TODO
    );
  }

  Future<VerificationKeyKesProduct> generateVerificationKey(
      SecretKeyKesProduct sk) async {
    final superTree = sk.superTree;
    if (superTree is KesMerkleNode) {
      return VerificationKeyKesProduct()
        ..value = (await kesHelper.witness(sk.superTree)).base58
        ..step = await getCurrentStep(sk);
    } else if (superTree is KesSigningLeaf) {
      return VerificationKeyKesProduct()
        ..value = (await kesHelper.witness(sk.superTree)).base58
        ..step = 0;
    } else {
      return VerificationKeyKesProduct()
        ..value = Uint8List(32).base58
        ..step = 0;
    }
  }

  _eraseLeafSecretKey(KesBinaryTree tree) {
    if (tree is KesMerkleNode) {
      if (tree.left is KesEmpty)
        return KesMerkleNode(tree.seed, tree.witnessLeft, tree.witnessRight,
            KesEmpty(), _eraseLeafSecretKey(tree.right));
      else if (tree.right is KesEmpty) {
        return KesMerkleNode(tree.seed, tree.witnessLeft, tree.witnessRight,
            _eraseLeafSecretKey(tree.left), KesEmpty());
      }
    } else if (tree is KesSigningLeaf) {
      kesHelper.overwriteBytes(tree.sk);
      return KesSigningLeaf(Uint8List(32), tree.vk);
    }
    throw Exception("Evolving Key Configuration Error");
  }
}

final _impl = KesProductImpl();

KesProduct kesProduct = KesProductImpl();

final _generateKeyPair = _impl.generateKeyPair;
final _getCurrentStep = _impl.getCurrentStep;
final _sign = _impl.sign;
final _update = _impl.update;
final _verify = _impl.verify;
final _generateVerificationKey = _impl.generateVerificationKey;

class TreeHeight {
  final int sup;
  final int sub;

  TreeHeight(this.sup, this.sub);

  @override
  int get hashCode => Object.hash(sup, sub);

  @override
  bool operator ==(Object other) {
    if (other is TreeHeight) {
      return sup == other.sup && sub == other.sub;
    }
    return false;
  }
}

class SecretKeyKesProduct {
  final KesBinaryTree superTree;
  final KesBinaryTree subTree;
  final List<int> nextSubSeed;
  final SignatureKesSum subSignature;
  final Int64 offset;

  SecretKeyKesProduct(
      {required this.superTree,
      required this.subTree,
      required this.nextSubSeed,
      required this.subSignature,
      required this.offset});

  @override
  int get hashCode =>
      Object.hash(superTree, subTree, nextSubSeed, subSignature, offset);

  @override
  bool operator ==(Object other) {
    if (other is SecretKeyKesProduct) {
      superTree == other.superTree &&
          subTree == other.subTree &&
          nextSubSeed.sameElements(other.nextSubSeed) &&
          subSignature == other.subSignature &&
          offset == other.offset;
    }
    return false;
  }

  factory SecretKeyKesProduct.decode(Uint8List bytes) {
    final superTreeRes = _decodeTree(bytes);
    final subTreeRes = _decodeTree(superTreeRes.$2);
    final nextSubSeed = subTreeRes.$2.sublist(0, 32);
    final subSignatureRes = _decodeSignature(subTreeRes.$2.sublist(32));
    final offset = Int64.fromBytesBigEndian(subSignatureRes.$2.sublist(0, 8));
    return SecretKeyKesProduct(
      superTree: superTreeRes.$1,
      subTree: subTreeRes.$1,
      nextSubSeed: nextSubSeed,
      subSignature: subSignatureRes.$1,
      offset: offset,
    );
  }

  List<int> get encode {
    return [
      ..._encodeTree(superTree),
      ..._encodeTree(subTree),
      ...nextSubSeed,
      ..._encodeSignature(subSignature),
      ...offset.toBytesBigEndian()
    ];
  }

  List<int> _encodeTree(KesBinaryTree tree) {
    if (tree is KesMerkleNode) {
      return [
        ...[0x00],
        ...tree.seed,
        ...tree.witnessLeft,
        ...tree.witnessRight,
        ..._encodeTree(tree.left),
        ..._encodeTree(tree.right)
      ];
    } else if (tree is KesSigningLeaf) {
      return [
        ...[0x01],
        ...tree.sk,
        ...tree.vk
      ];
    } else if (tree is KesEmpty) {
      return [0x02];
    }
    throw Exception("Encoding Error");
  }

  List<int> _encodeSignature(SignatureKesSum signature) {
    final result = [
      ...signature.verificationKey.decodeBase58,
      ...signature.signature.decodeBase58,
      ...Int64(signature.witness.length).toBytesBigEndian(),
    ];
    for (final t in signature.witness) result.addAll(t.decodeBase58);
    return result;
  }

  static (KesBinaryTree, Uint8List) _decodeTree(Uint8List bytes) {
    int cursor = 1;
    if (bytes[0] == 0x00) {
      final seed = bytes.sublist(cursor, cursor += 32);
      final witnessLeft = bytes.sublist(cursor, cursor += 32);
      final witnessRight = bytes.sublist(cursor, cursor += 32);
      final left = _decodeTree(bytes.sublist(cursor));
      final right = _decodeTree(left.$2);
      return (
        KesMerkleNode(seed, witnessLeft, witnessRight, left.$1, right.$1),
        right.$2
      );
    } else if (bytes[0] == 0x01) {
      final sk = bytes.sublist(cursor, cursor += 32);
      final vk = bytes.sublist(cursor, cursor += 32);
      return (KesSigningLeaf(sk, vk), bytes.sublist(cursor));
    } else if (bytes[0] == 0x02) {
      return (KesEmpty(), bytes.sublist(1));
    }
    throw Exception("Decoding Error");
  }

  static (SignatureKesSum, List<int>) _decodeSignature(List<int> bytes) {
    int cursor = 0;
    final vk = Uint8List.fromList(bytes.sublist(cursor, cursor += 32));
    final signature = Uint8List.fromList(bytes.sublist(cursor, cursor += 64));
    Int64 witnessLength = _parseInt(bytes.sublist(cursor, cursor += 8));
    final witness = List.generate(witnessLength.toInt(), (index) {
      return Uint8List.fromList(bytes.sublist(cursor, cursor += 32));
    });

    final kesSignature = SignatureKesSum()
      ..verificationKey = vk.base58
      ..signature = signature.base58
      ..witness.addAll(witness.map((e) => e.base58));
    return (kesSignature, bytes.sublist(cursor));
  }

  static Int64 _parseInt(List<int> bytes) {
    return Int64.fromBytesBigEndian(bytes);
  }
}

class KeyPairKesProduct {
  final SecretKeyKesProduct sk;
  final VerificationKeyKesProduct vk;

  KeyPairKesProduct({required this.sk, required this.vk});

  @override
  int get hashCode => Object.hash(sk, vk);

  @override
  bool operator ==(Object other) {
    return other is KeyPairKesProduct && other.sk == sk && other.vk == vk;
  }
}

class KesProudctIsolated extends KesProduct {
  final DComputeImpl _compute;

  KesProudctIsolated(this._compute);

  @override
  Future<KeyPairKesProduct> generateKeyPair(
          List<int> seed, TreeHeight height, Int64 offset) =>
      _compute((args) => _generateKeyPair(args.$1.$1, args.$1.$2, args.$2),
          ((seed, height), offset));

  @override
  Future<int> getCurrentStep(SecretKeyKesProduct sk) =>
      _compute(_getCurrentStep, sk);

  @override
  Future<SignatureKesProduct> sign(SecretKeyKesProduct sk, List<int> message) =>
      _compute((args) => _sign(args.$1, args.$2), (sk, message));

  @override
  Future<SecretKeyKesProduct> update(SecretKeyKesProduct sk, int step) =>
      _compute((args) => _update(args.$1, args.$2), (sk, step));

  @override
  Future<bool> verify(SignatureKesProduct signature, List<int> message,
          VerificationKeyKesProduct vk) =>
      _compute((args) => _verify(args.$1.$1, args.$1.$2, args.$2),
          ((signature, message), vk));

  @override
  Future<VerificationKeyKesProduct> generateVerificationKey(
          SecretKeyKesProduct sk) =>
      _compute(_generateVerificationKey, sk);
}

// class VerificationKeyKesProduct {
//   final Uint8List value;
//   final int step;

//   VerificationKeyKesProduct({required this.value, required this.step});

//   @override
//   int get hashCode => Object.hash(value, step);

//   @override
//   bool operator ==(Object other) {
//     if (other is VerificationKeyKesProduct) {
//       return value.sameElements(other.value) && step == other.step;
//     }
//     return false;
//   }
// }

// class SignatureKesProduct {
//   final SignatureKesSum superSignature;
//   final SignatureKesSum subSignature;
//   final Uint8List subRoot;

//   SignatureKesProduct({
//     required this.superSignature,
//     required this.subSignature,
//     required this.subRoot,
//   });

//   @override
//   int get hashCode => Object.hash(superSignature, subSignature, subRoot);

//   @override
//   bool operator ==(Object other) {
//     if (other is SignatureKesProduct) {
//       return superSignature == other.superSignature &&
//           subSignature == other.subSignature &&
//           subRoot.sameElements(other.subRoot);
//     }
//     return false;
//   }
// }
